server/remote_asset/fetch_server/fetch_server.go

```diff
@@ -203,11 +186,70 @@ func (p *FetchServer) FetchDirectory(ctx context.Context, req *rapb.FetchDirecto
 	return nil, status.UnimplementedError("FetchDirectory is not yet implemented")
 }
 
+func (p *FetchServer) findBlobInCache(ctx context.Context, instanceName string, checksumFunc repb.DigestFunction_Value, checksum string, storageFunc repb.DigestFunction_Value) *repb.Digest {
+	blobDigest := &repb.Digest{
+		Hash: checksum,
+		// The digest size is unknown since the client only sends up
+		// the hash. We can look up the size using the Metadata API,
+		// which looks up only using the hash, so the size we pass here
+		// doesn't matter.
+		SizeBytes: 1,
+	}
+	expectedHash := blobDigest.Hash
+	cacheRN := digest.NewResourceName(blobDigest, instanceName, rspb.CacheType_CAS, checksumFunc)
+	log.CtxInfof(ctx, "Looking up %s in cache", blobDigest.Hash)
+
+	// Lookup metadata to get the correct digest size to be returned to
+	// the client.
+	cache := p.env.GetCache()
+	md, err := cache.Metadata(ctx, cacheRN.ToProto())
+	if err != nil {
+		log.CtxInfof(ctx, "FetchServer failed to get metadata for %s: %s", expectedHash, err)
+		return nil
+	}
+	blobDigest.SizeBytes = md.DigestSizeBytes
+
+	// Even though we successfully fetched metadata, we need to renew
+	// the cache entry (using Contains()) to ensure that it doesn't
+	// expire by the time the client requests it from cache.
+	cacheRN = digest.NewResourceName(blobDigest, instanceName, rspb.CacheType_CAS, checksumFunc)
+	exists, err := cache.Contains(ctx, cacheRN.ToProto())
+	if err != nil {
+		log.CtxErrorf(ctx, "Failed to renew %s: %s", digest.String(blobDigest), err)
+		return nil
+	}
+	if !exists {
+		log.CtxInfof(ctx, "Blob %s expired before we could renew it", digest.String(blobDigest))
+		return nil
+	}
+
+	// If the digestFunc is supplied and differ from the checksum sri,
```

Comment:
I feel like it's a bit non-intuitive to be doing an upload in `findBlobInCache` - maybe split this into a separate func?

Reply:
Moved to a separate func.